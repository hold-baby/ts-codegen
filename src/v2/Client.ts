import {
  BodyParameter,
  FormDataParameter,
  Operation,
  Parameter,
  Path,
  PathParameter,
  QueryParameter,
  Reference,
  Response,
} from "swagger-schema-official";
import { compact, Dictionary, filter, get, isEmpty, keys, map, mapValues, pick, reduce, sortBy } from "lodash";
import { getRequestURL, setDeprecated, toCapitalCase, toTypes } from "src/core/utils";
import { CustomType, Ref, Register } from "src/core/Type";
import { Schema } from "src/core/Schema";

type TPaths = { [pathName: string]: Path };

interface IClientConfig {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  bodyParams: string[];
  formDataParams: string[];
  deprecated?: boolean;
}

export class Client {
  clientConfigs: IClientConfig[] = [];
  schemaHandler: Schema;

  static of(paths: TPaths, basePath: string = "") {
    return new Client(paths, basePath);
  }

  constructor(private paths: TPaths, private basePath: string) {
    this.schemaHandler = new Schema();
    this.clientConfigs = reduce(
      this.paths,
      (configs: IClientConfig[], path: Path, pathName: string) => [...configs, ...this.buildConfig(path, pathName)],
      [],
    );
  }

  buildConfig(path: Path, pathName: string) {
    const operations = pick(path, ["get", "post", "put", "delete", "patch", "head"]);

    return keys(operations).map((method) => {
      const operation = (operations as Dictionary<any>)[method];
      const pickParamsByType = pickParams(operation.parameters as Parameter[]);
      const params = {
        pathParams: pickParamsByType("path") as PathParameter[],
        queryParams: pickParamsByType("query") as QueryParameter[],
        bodyParams: pickParamsByType("body") as BodyParameter[],
        formDataParams: pickParamsByType("formData") as FormDataParameter[],
      };

      return {
        url: getRequestURL(pathName, this.basePath),
        method,
        operationId: operation.operationId,
        TResp: this.getResponseType(operation.responses),
        TReq: {
          ...this.getParamTypes(params.pathParams, operation.operationId),
          ...this.getParamTypes(params.queryParams, operation.operationId),
          ...this.getParamTypes(params.bodyParams, operation.operationId),
          ...this.getParamTypes(params.formDataParams, operation.operationId),
        },
        pathParams: getParamsNames(params.pathParams),
        queryParams: getParamsNames(params.queryParams),
        bodyParams: getParamsNames(params.bodyParams),
        formDataParams: getParamsNames(params.formDataParams),
        deprecated: operation.deprecated,
      };
    });
  }

  getParamTypes = (
    params: Array<PathParameter | BodyParameter | QueryParameter | FormDataParameter>,
    _name: string,
  ): { [key: string]: CustomType } => {
    return params.reduce(
      (results, param) => ({
        ...results,
        [propName(param)]: this.schemaHandler.convert(
          get(param, "schema", param),
          `${toCapitalCase(_name)}${toCapitalCase(param.name)}`,
        ),
      }),
      {},
    );
  };

  getResponseType = (responses: Operation["responses"]) => {
    const response200 = get(responses, "200");
    const response201 = get(responses, "201");

    if ((response200 as Reference)?.$ref || (response201 as Reference)?.$ref) {
      return this.schemaHandler.convert(response200 || response201);
    }

    const v = (response200 as Response)?.schema || (response201 as Response)?.schema;
    return v ? this.schemaHandler.convert(v) : v;
  };
}

const pickParams = (parameters: Array<Parameter | Reference>) => (type: "path" | "query" | "body" | "formData") =>
  filter(parameters, (param) => (param as Parameter).in === type);

const getParamsNames = (params: any[]) => (isEmpty(params) ? [] : map(params, (param) => (param as Parameter).name));

const toRequestParams = (data: any[] = []) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;
const propName = (param: Parameter) => `${param.name}${param.required ? "" : "?"}`;

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const toRequest = (config: IClientConfig): string[] => {
  for (let name in Register.refs) {
    if (!(Register.refs[name] as Ref).alias) {
      (Register.refs[name] as Ref).rename(addPrefix(name));
    }
  }

  const clientConfig = sortBy(config, (o) => o.operationId);
  const requests = clientConfig.map((v: IClientConfig) => {
    const TReq = !isEmpty(v.TReq) ? toTypes(mapValues(v.TReq, (v) => v.toType())) : "";
    const getRequestBody = () => {
      if (isEmpty(v.bodyParams) && isEmpty(v.formDataParams)) {
        return null;
      }
      if (isEmpty(v.bodyParams)) {
        return v.formDataParams;
      }
      return v.bodyParams;
    };

    const getData = () => {
      const requestBody = getRequestBody();
      if (!requestBody) {
        return "";
      }
      return requestBody.length > 1 ? `data: {${requestBody.join(",")}},` : `data: ${requestBody},`;
    };

    const requestParamList = compact([...v.pathParams, ...v.queryParams, ...v.bodyParams, ...v.formDataParams]);
    const requestInputs = isEmpty(requestParamList) ? "" : toRequestParams(requestParamList);

    const getParams = () => {
      const params = toRequestParams(get(v, "queryParams"));
      return params ? `params: ${params},` : "";
    };

    const getHeaders = () => {
      const requestBody = getRequestBody();
      if (!requestBody) {
        return "";
      }
      return `headers: { "Content-Type": ${
        !isEmpty(v.formDataParams) ? "'multipart/form-data'" : "'application/json'"
      } },`;
    };

    return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction<${TReq}, ${v.TResp?.toType() || ""}>("${
      v.operationId
    }", (${requestInputs}) => ({ url: \`${v.url}\`,method: "${v.method}",${getData()}${getParams()}${getHeaders()} })
);
`;
  });

  return requests;
};
