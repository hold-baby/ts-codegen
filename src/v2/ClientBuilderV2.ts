import {
  BodyParameter,
  FormDataParameter,
  Operation,
  Parameter,
  Path,
  PathParameter,
  QueryParameter,
  Reference,
  Response,
} from "swagger-schema-official";
import { Dictionary, filter, get, isEmpty, keys, map, pick, reduce, sortBy } from "lodash";
import { generateEnums, getRequestURL, setDeprecated, toTypes } from "src/core/utils";
import { SchemaResolver } from "src/core/SchemaResolver";

type TPaths = { [pathName: string]: Path };

interface IClientConfig {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  bodyParams: string[];
  formDataParams: string[];
  deprecated?: boolean;
}

const pickParams = (parameters: Array<Parameter | Reference>) => (type: "path" | "query" | "body" | "formData") =>
  filter(parameters, (param) => (param as Parameter).in === type);

const getParamsNames = (params: any[]) => (isEmpty(params) ? [] : map(params, (param) => (param as Parameter).name));

const propName = (param: Parameter) => `${param.name}${param.required ? "" : "?"}`;

export class ClientBuilderV2 {
  schemaResolver: SchemaResolver;
  clientConfig: IClientConfig[] = [];
  enums: Dictionary<any> = {};

  static of(paths: TPaths, basePath: string = "") {
    return new ClientBuilderV2(paths, basePath);
  }

  constructor(private paths: TPaths, private basePath: string) {
    this.schemaResolver = SchemaResolver.of((k, v) => {
      if (k) {
        this.enums[k] = v;
      }
    });
  }

  toRequest = (): string[] => {
    const data = sortBy(this.clientConfig, (o) => o.operationId);
    const requests = data.map((v: IClientConfig) => {
      const TReq = !isEmpty(v.TReq) ? toTypes(v.TReq) : undefined;
      const requestParamList = [...v.pathParams, ...v.queryParams, ...v.bodyParams, ...v.formDataParams];
      const bodyData = v.bodyParams;
      const formData = v.formDataParams;
      const body = !isEmpty(bodyData) ? bodyData : formData;
      const params = this.toRequestParams(get(v, "queryParams"));
      return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction<${TReq}, ${v.TResp}>('${v.operationId}', (${
        !isEmpty(requestParamList) ? `${this.toRequestParams(requestParamList)}` : ""
      }) => ({url: \`${v.url}\`, method: "${v.method}", ${
        !isEmpty(body) ? (body.length > 1 ? `data: {${body.join(",")}},` : `data: ${body},`) : ""
      }${params ? `params: ${params},` : ""}${
        body ? `headers: {'Content-Type': ${!isEmpty(formData) ? "'multipart/form-data'" : "'application/json'"}}` : ""
      }}));`;
    });

    const enums = keys(this.enums).map((k) => generateEnums(this.enums, k));
    return [...requests, ...enums];
  };

  scan = () => {
    this.clientConfig = reduce(
      this.paths,
      (config: IClientConfig[], path: Path, pathName: string) => [...config, ...this.buildConfig(path, pathName)],
      [],
    );
    return this;
  };

  buildConfig(path: Path, pathName: string) {
    const operations = pick(path, ["get", "post", "put", "delete", "patch", "head"]);

    return keys(operations).map((method) => {
      const operation = (operations as Dictionary<any>)[method];
      const pickParamsByType = pickParams(operation.parameters as Parameter[]);
      const params = {
        pathParams: pickParamsByType("path") as PathParameter[],
        queryParams: pickParamsByType("query") as QueryParameter[],
        bodyParams: pickParamsByType("body") as BodyParameter[],
        formDataParams: pickParamsByType("formData") as FormDataParameter[],
      };

      return {
        url: getRequestURL(pathName, this.basePath),
        method,
        operationId: operation.operationId,
        TResp: this.getResponseType(operation.responses),
        TReq: {
          ...this.getParamTypes(params.pathParams),
          ...this.getParamTypes(params.queryParams),
          ...this.getParamTypes(params.bodyParams),
          ...this.getParamTypes(params.formDataParams),
        },
        pathParams: getParamsNames(params.pathParams),
        queryParams: getParamsNames(params.queryParams),
        bodyParams: getParamsNames(params.bodyParams),
        formDataParams: getParamsNames(params.formDataParams),
        deprecated: operation.deprecated,
      };
    });
  }

  toRequestParams = (data: any[] = []) =>
    !isEmpty(data)
      ? `{
    ${data.join(",\n")}
    }`
      : undefined;

  getParamTypes = (params: Array<PathParameter | BodyParameter | QueryParameter | FormDataParameter>) =>
    params.reduce(
      (results, param) => ({
        ...results,
        [propName(param)]: this.schemaResolver.toType({
          ...((param as any).schema ? (param as any).schema : param),
          _name: param.name,
          _propKey: param.name,
        }),
      }),
      {},
    );

  getResponseType = (responses: Operation["responses"]) => {
    const response200 = get(responses, "200");
    const response201 = get(responses, "201");

    if ((response200 as Reference)?.$ref || (response201 as Reference)?.$ref) {
      return this.schemaResolver.toType(response200 || response201);
    }

    return this.schemaResolver.toType((response200 as Response)?.schema || (response201 as Response)?.schema);
  };
}
