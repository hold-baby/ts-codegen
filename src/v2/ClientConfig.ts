import {
  BodyParameter,
  FormDataParameter,
  Operation,
  Parameter,
  Path,
  PathParameter,
  QueryParameter,
  Reference,
  Response,
} from "swagger-schema-official";
import { compact, filter, get, isEmpty, keys, map, mapValues, pick, reduce, sortBy } from "lodash";
import { getRequestURL, setDeprecated, toCapitalCase, toTypes } from "src/core/utils";
import { CustomType, Ref, Register } from "src/core/Type";
import { Schema } from "src/core/Schema";

type Paths = { [pathName: string]: Path };

interface IClientConfig {
  url: string;
  method: string;
  TResp: CustomType | undefined;
  TReq: { [key: string]: CustomType } | undefined;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  contentType: string;
  deprecated?: boolean;
}

export class ClientConfig {
  clientConfigs: IClientConfig[] = [];
  schemaHandler: Schema;

  static of(paths: Paths, basePath: string = "") {
    return new ClientConfig(paths, basePath);
  }

  constructor(private paths: Paths, private basePath: string) {
    this.schemaHandler = new Schema();
    this.clientConfigs = reduce(
      this.paths,
      (configs: IClientConfig[], path: Path, pathName: string) => [...configs, ...this.buildConfig(path, pathName)],
      [],
    );
  }

  buildConfig(path: Path, pathName: string) {
    // TODO: handle head method later
    const operations = pick(path, ["get", "post", "put", "delete", "patch", "head"]) as { [method: string]: Operation };

    return keys(operations).map((method) => {
      const operation = operations[method];
      const pickParamsByType = pickParams(operation.parameters as Parameter[]);
      const pathParams = pickParamsByType("path") as PathParameter[];
      const queryParams = pickParamsByType("query") as QueryParameter[];
      const bodyParams = pickParamsByType("body") as BodyParameter[];
      const formDataParams = pickParamsByType("formData") as FormDataParameter[];
      const getParamTypes = this.getParamTypes(operation.operationId);

      return {
        url: getRequestURL(pathName, this.basePath),
        method,
        operationId: operation.operationId, //camelCase(operation.operationId) TODO: add tests later, 向后兼容？
        TResp: this.getResponseType(operation.responses),
        TReq: {
          ...getParamTypes(pathParams),
          ...getParamTypes(queryParams),
          ...getParamTypes(bodyParams),
          ...getParamTypes(formDataParams),
        },
        pathParams: getParamsNames(pathParams),
        queryParams: getParamsNames(queryParams),
        contentType: this.getContentType(bodyParams, formDataParams),
        deprecated: operation.deprecated,
      };
    });
  }

  getContentType(bodyParams: BodyParameter[], formData: FormDataParameter[]) {
    if (!isEmpty(bodyParams)) {
      return "application/json";
    }
    if (!isEmpty(formData)) {
      return "multipart/form-data";
    }
    return "";
  }

  getParamTypes = (_name?: string) => {
    return (
      params: Array<PathParameter | BodyParameter | QueryParameter | FormDataParameter>,
    ): { [key: string]: CustomType } => {
      return params.reduce(
        (results, param) => ({
          ...results,
          [propName(param)]: this.schemaHandler.convert(
            get(param, "schema", param),
            `${toCapitalCase(_name)}${toCapitalCase(param.name)}`,
          ),
        }),
        {},
      );
    };
  };

  getResponseType = (responses: Operation["responses"]) => {
    const response200 = get(responses, "200");
    const response201 = get(responses, "201");

    if ((response200 as Reference)?.$ref || (response201 as Reference)?.$ref) {
      return this.schemaHandler.convert(response200 || response201);
    }

    const v = (response200 as Response)?.schema || (response201 as Response)?.schema;
    return v ? this.schemaHandler.convert(v) : undefined;
  };
}

const pickParams = (parameters: Array<Parameter | Reference>) => (type: "path" | "query" | "body" | "formData") =>
  filter(parameters, (param) => (param as Parameter).in === type);

const getParamsNames = (params: any[]) => (isEmpty(params) ? [] : map(params, (param) => (param as Parameter).name));

const toRequestParams = (data: any[] = []) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;
const propName = (param: Parameter) => `${param.name}${param.required ? "" : "?"}`;

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const toRequest = (clientConfigs: IClientConfig[]): string[] => {
  for (let name in Register.refs) {
    if (!(Register.refs[name] as Ref).alias) {
      (Register.refs[name] as Ref).rename(addPrefix(name));
    }
  }
  const clientConfig = sortBy(clientConfigs, (o) => o.operationId);

  return clientConfig.map((v: IClientConfig) => {
    const TReq = !isEmpty(v.TReq) ? toTypes(mapValues(v.TReq, (v) => v.toType())) : "";
    const requestParamList = compact([...v.pathParams, ...v.queryParams, v.contentType ? "requestBody" : ""]);
    const requestInputs = isEmpty(requestParamList) ? "" : toRequestParams(requestParamList);

    const getParams = () => {
      const params = toRequestParams(get(v, "queryParams"));
      return params ? `params: ${params},` : "";
    };

    const getHeaders = () => (v.contentType ? `headers: { "Content-Type": '${v.contentType}' },` : "");

    return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction<${TReq}, ${v.TResp?.toType() || ""}>("${
      v.operationId
    }", (${requestInputs}) => ({ url: \`${v.url}\`,method: "${v.method}",${
      v.contentType ? `data: requestBody,` : ""
    }${getParams()}${getHeaders()} })
);
`;
  });
};
