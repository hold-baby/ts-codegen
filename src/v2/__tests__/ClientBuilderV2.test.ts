import { ClientBuilderV2 } from "src/v2/ClientBuilderV2";
import swaggerV2 from "examples/swagger.v2.json";

describe("PathResolver", () => {
  it("should get resolved paths by swagger schema", () => {
    expect(ClientBuilderV2.of((swaggerV2 as any).paths, swaggerV2.basePath).scan().resolvedPaths).toEqual(
      expectedPathResolvedData,
    );
  });

  it("should get correct action creator by resolved paths", () => {
    expect(
      ClientBuilderV2.of((swaggerV2 as any).paths, swaggerV2.basePath)
        .scan()
        .toRequest(),
    ).toEqual(expectedRequest);
  });
});

const expectedPathResolvedData = [
  {
    TReq: { attachment: "File" },
    TResp: "IAttachmentBo",
    bodyParams: [],
    formDataParams: ["attachment"],
    operationId: "uploadAttachmentUsingPOST",
    pathParams: [],
    queryParams: [],
    url: "/api/test",
    method: "post",
  },
  {
    TReq: { id: "string" },
    TResp: "IResource",
    bodyParams: [],
    formDataParams: [],
    operationId: "downloadUsingGET",
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/${id}",
    method: "get",
  },
  {
    TReq: { id: "string" },
    TResp: "",
    bodyParams: [],
    formDataParams: [],
    operationId: "deleteAttachmentUsingDELETE",
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/${id}",
    method: "delete",
  },
  {
    TReq: { id: "string" },
    TResp: "IBookDetailVo",
    bodyParams: [],
    formDataParams: [],
    operationId: "findBookByIdUsingGET",
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/book/${id}",
    method: "get",
  },
  {
    TReq: { id: "string", updateBookRequest: "IUpdateBookRequest" },
    TResp: "",
    bodyParams: ["updateBookRequest"],
    formDataParams: [],
    operationId: "updateBookByIdUsingPUT",
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/book/${id}",
    method: "put",
  },
  {
    TReq: { scheduleDate: "number", "roleId?": "string" },
    TResp: "IScheduleVo[]",
    bodyParams: [],
    formDataParams: [],
    operationId: "getScheduleDetailsByDateUsingGET",
    pathParams: [],
    queryParams: ["scheduleDate", "roleId"],
    url: "/api/test/schedules",
    method: "get",
  },
  {
    TReq: { "from?": "keyof typeof FromFrom#EnumSuffix", documentId: "string" },
    TResp: "IDocumentVo",
    bodyParams: [],
    formDataParams: [],
    operationId: "getDocumentByIdUsingGET",
    pathParams: ["documentId"],
    queryParams: ["from"],
    url: "/api/test/documents/${documentId}/doc",
    method: "get",
  },
];

const expectedRequest = [
  "export const deleteAttachmentUsingDELETE = createRequestAction<{\n  " +
    "      'id': string;\n      }, >('deleteAttachmentUsingDELETE', ({\n  " +
    '  id\n    }) => ({url: `/api/test/${id}`, method: "delete", }));',
  "export const downloadUsingGET = createRequestAction<{\n        " +
    "'id': string;\n      }, IResource>('downloadUsingGET', ({\n    " +
    'id\n    }) => ({url: `/api/test/${id}`, method: "get", }));',
  "export const findBookByIdUsingGET = createRequestAction<{\n        " +
    "'id': string;\n      }, IBookDetailVo>('findBookByIdUsingGET', ({\n  " +
    '  id\n    }) => ({url: `/api/test/book/${id}`, method: "get", }));',
  "export const getDocumentByIdUsingGET = createRequestAction<{\n" +
    "        'documentId': string;\n'from'?: keyof typeof " +
    "FromFrom;\n      }, IDocumentVo>('getDocumentByIdUsingGET', " +
    "({\n    documentId,\nfrom\n    }) => ({url: " +
    '`/api/test/documents/${documentId}/doc`, method: "get", ' +
    "params: {\n    from\n    },}));",
  "export const getScheduleDetailsByDateUsingGET = " +
    "createRequestAction<{\n        'roleId'?: string;\n" +
    "'scheduleDate': number;\n      }, " +
    "IScheduleVo[]>('getScheduleDetailsByDateUsingGET', ({\n    " +
    "scheduleDate,\nroleId\n    }) => ({url: `/api/test/schedules`, " +
    'method: "get", params: {\n    scheduleDate,\nroleId\n    },}));',
  "export const updateBookByIdUsingPUT = createRequestAction<{\n     " +
    "   'id': string;\n'updateBookRequest': IUpdateBookRequest;\n      " +
    "}, >('updateBookByIdUsingPUT', ({\n    id,\nupdateBookRequest\n    " +
    '}) => ({url: `/api/test/book/${id}`, method: "put", data: ' +
    "updateBookRequest,headers: {'Content-Type': " +
    "'application/json'}}));",
  "export const uploadAttachmentUsingPOST = createRequestAction<{\n        " +
    "'attachment': File;\n      }, " +
    "IAttachmentBo>('uploadAttachmentUsingPOST', ({\n    attachment\n    }) " +
    '=> ({url: `/api/test`, method: "post", data: attachment,headers: ' +
    "{'Content-Type': 'multipart/form-data'}}));",
  'export enum FromFrom {"AAA"="AAA","BBB"="BBB"}',
];
