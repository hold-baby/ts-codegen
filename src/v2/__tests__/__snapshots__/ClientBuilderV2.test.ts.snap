// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathResolver should get correct action creator by resolved paths 1`] = `
"export const deleteAttachmentUsingDELETE = createRequestAction<{
  id: string;
}>(\\"deleteAttachmentUsingDELETE\\", ({ id }) => ({
  url: \`/api/test/\${id}\`,
  method: \\"delete\\",
}));

export const downloadUsingGET = createRequestAction<
  {
    id: string;
  },
  IResource
>(\\"downloadUsingGET\\", ({ id }) => ({
  url: \`/api/test/\${id}\`,
  method: \\"get\\",
}));

export const findBookByIdUsingGET = createRequestAction<
  {
    id: string;
  },
  IBookDetailVo
>(\\"findBookByIdUsingGET\\", ({ id }) => ({
  url: \`/api/test/book/\${id}\`,
  method: \\"get\\",
}));

export const getDocumentByIdUsingGET = createRequestAction<
  {
    documentId: string;
    from?: keyof typeof FromFrom;
  },
  IDocumentVo
>(\\"getDocumentByIdUsingGET\\", ({ documentId, from }) => ({
  url: \`/api/test/documents/\${documentId}/doc\`,
  method: \\"get\\",
  params: { from },
}));

export const getScheduleDetailsByDateUsingGET = createRequestAction<
  {
    roleId?: string;
    scheduleDate: number;
  },
  IScheduleVo[]
>(\\"getScheduleDetailsByDateUsingGET\\", ({ scheduleDate, roleId }) => ({
  url: \`/api/test/schedules\`,
  method: \\"get\\",
  params: { scheduleDate, roleId },
}));

export const updateBookByIdUsingPUT = createRequestAction<{
  id: string;
  updateBookRequest: IUpdateBookRequest;
}>(\\"updateBookByIdUsingPUT\\", ({ id, updateBookRequest }) => ({
  url: \`/api/test/book/\${id}\`,
  method: \\"put\\",
  data: updateBookRequest,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadAttachmentUsingPOST = createRequestAction<
  {
    attachment: File;
  },
  IAttachmentBo
>(\\"uploadAttachmentUsingPOST\\", ({ attachment }) => ({
  url: \`/api/test\`,
  method: \\"post\\",
  data: attachment,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));
export enum FromFrom {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}
"
`;

exports[`PathResolver should handle multiple form data in response 1`] = `
"export const addPet = createRequestAction<{
  body: IPet;
}>(\\"addPet\\", ({ body }) => ({
  url: \`/api/test/pet\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUser = createRequestAction<{
  body: IUser;
}>(\\"createUser\\", ({ body }) => ({
  url: \`/api/test/user\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithArrayInput = createRequestAction<{
  body: IUser[];
}>(\\"createUsersWithArrayInput\\", ({ body }) => ({
  url: \`/api/test/user/createWithArray\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithListInput = createRequestAction<{
  body: IUser[];
}>(\\"createUsersWithListInput\\", ({ body }) => ({
  url: \`/api/test/user/createWithList\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deleteOrder = createRequestAction<{
  orderId: number;
}>(\\"deleteOrder\\", ({ orderId }) => ({
  url: \`/api/test/store/order/\${orderId}\`,
  method: \\"delete\\",
}));

export const deletePet = createRequestAction<{
  petId: number;
}>(\\"deletePet\\", ({ petId }) => ({
  url: \`/api/test/pet/\${petId}\`,
  method: \\"delete\\",
}));

export const deleteUser = createRequestAction<{
  username: string;
}>(\\"deleteUser\\", ({ username }) => ({
  url: \`/api/test/user/\${username}\`,
  method: \\"delete\\",
}));

export const findPetsByStatus = createRequestAction<
  {
    status: keyof typeof StatusStatus[];
  },
  IPet[]
>(\\"findPetsByStatus\\", ({ status }) => ({
  url: \`/api/test/pet/findByStatus\`,
  method: \\"get\\",
  params: { status },
}));

/**
 * @deprecated findPetsByTags
 */

export const findPetsByTags = createRequestAction<
  {
    tags: string[];
  },
  IPet[]
>(\\"findPetsByTags\\", ({ tags }) => ({
  url: \`/api/test/pet/findByTags\`,
  method: \\"get\\",
  params: { tags },
}));

export const getInventory = createRequestAction<undefined, object>(\\"getInventory\\", () => ({
  url: \`/api/test/store/inventory\`,
  method: \\"get\\",
}));

export const getOrderById = createRequestAction<
  {
    orderId: number;
  },
  IOrder
>(\\"getOrderById\\", ({ orderId }) => ({
  url: \`/api/test/store/order/\${orderId}\`,
  method: \\"get\\",
}));

export const getPetById = createRequestAction<
  {
    petId: number;
  },
  IPet
>(\\"getPetById\\", ({ petId }) => ({
  url: \`/api/test/pet/\${petId}\`,
  method: \\"get\\",
}));

export const getUserByName = createRequestAction<
  {
    username: string;
  },
  IUser
>(\\"getUserByName\\", ({ username }) => ({
  url: \`/api/test/user/\${username}\`,
  method: \\"get\\",
}));

export const loginUser = createRequestAction<
  {
    password: string;
    username: string;
  },
  string
>(\\"loginUser\\", ({ username, password }) => ({
  url: \`/api/test/user/login\`,
  method: \\"get\\",
  params: { username, password },
}));

export const logoutUser = createRequestAction<undefined>(\\"logoutUser\\", () => ({
  url: \`/api/test/user/logout\`,
  method: \\"get\\",
}));

export const placeOrder = createRequestAction<
  {
    body: IOrder;
  },
  IOrder
>(\\"placeOrder\\", ({ body }) => ({
  url: \`/api/test/store/order\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequestAction<{
  body: IPet;
}>(\\"updatePet\\", ({ body }) => ({
  url: \`/api/test/pet\`,
  method: \\"put\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePetWithForm = createRequestAction<{
  name?: string;
  petId: number;
  status?: string;
}>(\\"updatePetWithForm\\", ({ petId, name, status }) => ({
  url: \`/api/test/pet/\${petId}\`,
  method: \\"post\\",
  data: { name, status },
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const updateUser = createRequestAction<{
  body: IUser;
  username: string;
}>(\\"updateUser\\", ({ username, body }) => ({
  url: \`/api/test/user/\${username}\`,
  method: \\"put\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadFile = createRequestAction<
  {
    additionalMetadata?: string;
    file?: File;
    petId: number;
  },
  IApiResponse
>(\\"uploadFile\\", ({ petId, additionalMetadata, file }) => ({
  url: \`/api/test/pet/\${petId}/uploadImage\`,
  method: \\"post\\",
  data: { additionalMetadata, file },
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));
export enum StatusStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}
"
`;
